rllib/algorithms/algorithm.py

```diff
@@ -1503,60 +1522,72 @@ def training_step(self) -> ResultDict:
+                    worker_set=self.workers,
+                    max_env_steps=self.config.train_batch_size,
+                    sample_timeout_s=self.config.sample_timeout_s,
+                    _uses_new_env_runners=self.config.uses_new_env_runners,
+                    _return_metrics=True,
                 )
-
         train_batch = train_batch.as_multi_agent()
-        self._counters[NUM_AGENT_STEPS_SAMPLED] += train_batch.agent_steps()
-        self._counters[NUM_ENV_STEPS_SAMPLED] += train_batch.env_steps()
+
+        # Reduce EnvRunner metrics over the n EnvRunners.
+        self.metrics.log_n_stats_dicts(
+            key=ENV_RUNNER_RESULTS, stats_dicts=env_runner_metrics
+        )
+        #self._counters[NUM_AGENT_STEPS_SAMPLED] += train_batch.agent_steps()
+        #self._counters[NUM_ENV_STEPS_SAMPLED] += train_batch.env_steps()
 
         # Only train if train_batch is not empty.
         # In an extreme situation, all rollout workers die during the
         # synchronous_parallel_sample() call above.
         # In which case, we should skip training, wait a little bit, then probe again.
-        train_results = {}
-        if train_batch.agent_steps() > 0:
-            # Use simple optimizer (only for multi-agent or tf-eager; all other
-            # cases should use the multi-GPU optimizer, even if only using 1 GPU).
-            # TODO: (sven) rename MultiGPUOptimizer into something more
-            #  meaningful.
-            if self.config._enable_new_api_stack:
-                train_results = self.learner_group.update_from_batch(batch=train_batch)
-            elif self.config.get("simple_optimizer") is True:
-                train_results = train_one_step(self, train_batch)
+        with self.metrics.log_time(LEARNER_UPDATE_TIMER):
+            if train_batch.agent_steps() > 0:
+                learner_results = (
+                    self.learner_group.update_from_batch(batch=train_batch)
+                )
+                self.metrics.log_stats_dict(
+                    key=LEARNER_RESULTS, stats_dict=learner_results
+                )
             else:
-                train_results = multi_gpu_train_one_step(self, train_batch)
-        else:
-            # Wait 1 sec before probing again via weight syncing.
-            time.sleep(1.0)
-
-        # Update weights and global_vars - after learning on the local worker - on all
-        # remote workers (only those policies that were actually trained).
-        global_vars = {
-            "timestep": self._counters[NUM_ENV_STEPS_SAMPLED],
-        }
-        with self._timers[SYNCH_WORKER_WEIGHTS_TIMER]:
-            # TODO (Avnish): Implement this on learner_group.get_weights().
-            from_worker_or_trainer = None
-            if self.config._enable_new_api_stack:
-                from_worker_or_trainer = self.learner_group
+                # Wait 1 sec before probing again via weight syncing.
+                time.sleep(1.0)
```

Comment:
True, I don't even know why we would sleep here at all. I guess to avoid CPU spinning (for no good reason, as the env runners are all down in this situation).

Reduced to 0.01 sec.

Reply:
Keep in mind, that none of our algos uses this code. It's here for demonstration purposes only.